### ISSUE TITLE
[cssom-1] How to serialize specified value of properties with a var()?

### ISSUE THREAD

**Issue by tabatkins** [w3c/csswg-drafts#6484]
(continuing topic from <https://github.com/w3c/csswg-drafts/issues/5685>)

It's not currently well-specified how serialization for specified values of non-custom properties should behave when they have a var() reference.

That's a lot of words! I'm referring to [this example](http://software.hixie.ch/utilities/js/live-dom-viewer/saved/9522):

```html
<!DOCTYPE html>
<body style="color: /* foo */ var(--foo); --foo: blue;">
<script>
document.body.textContent = document.body.style.getPropertyValue("color");
</script>
```

Currently, Chrome serializes this as `var(--bar)`, while Firefox serializes it as `/* foo */ var(--bar)`. (I'm not sure what WebKit does.)

All browser, when given a property *without* a variable reference, like `color: /* foo */ red;`, serialize the specified value as just the keyword, without the comments.

In the f2f earlier today, @emilio argued that Firefox's behavior is because it holds onto the original string for properties with variable references, and serializing using it is simplest. I *think* that Chrome's behavior is because it stores props with var() references as a token stream (throwing away comments), and just reserializes from the token stream.

**Comment by emilio** [comment]
Another interesting test-case is something not involving comments like:

```
<!DOCTYPE html>
<body style="color: 1e3 var(--foo); --foo: blue;">
<script>
document.body.textContent = document.body.style.getPropertyValue("color");
</script>
```

Chrome serializes `1000 var(--foo)` while Firefox serializes `1e3 var(--foo)`.

**Comment by tabatkins** [comment]
Yeah that follows from the "parse as token stream then serialize directly from the tokens" strategy.

**Comment by andruud** [comment]
> I think that Chrome's behavior is because it stores props with var() references as a token stream (throwing away comments)

Yeah, that's right.

Since we (should) serialize `--color: /* foo */ var(--foo)` as the original string, it seems consistent to serialize `color: /* foo */ var(--foo)` as the original string as well.

**Comment by tabatkins** [comment]
I'm fine with specifying that, it just needed to be made clear. If you're willing to make the impl change in Chrome, then Firefox can stay as-is, and @emilio and I can figure out between the two of us whether this requires spec edits to CSSOM or Variables. ^_^

**Comment by andruud** [comment]
cc @kbabbitt

**Comment by cdoublev** [comment]
I notice that Firefox trims leading whitespace(s) of the original string and that other consecutive whitespace(s) are kept, (consecutive whitespaces are replaced by `__` in the following example, because the Markdown parser merges them) eg. `color:__var(__--blue__)__` becomes `color: var(__--blue__)__`. I know that leading whitespace(s) are trimmed when tokenizing but I do not know a procedure in CSSOM, CSS Syntax, or CSS Variable, that defines trimming leading whitespaces (before tokenizing), though.

I understand that your intent is to define that the original string (and its leading whitespaces?) should be kept.

To serialize a CSS value, the declaration value should be *represented as a list of CSS component values `components` that, when parsed according to the property’s grammar, would represent that value*, and a component value is defined as *one of the preserved tokens, a function, or a simple block*. Does the intended change(s) in the spec(s) mean that a CSS component value can be a string in case the specified value is a `var()`-containing value?

**Comment by cdoublev** [comment]
> Right, as per those issues you linked in your edit, leading (and trailing) whitespace is trimmed by the Syntax spec [...]. So Firefox is correct.

Then the trailing whitespace should be removed when serializing `color:__var(__--blue__)__`, isn't it? The output in Firefox is `color: var(__--blue__)__`.

> > Does the intended change(s) in the spec(s) mean that a CSS component value can be a string in case the specified value is a var()-containing value?
>
> I'm sorry, I'm not quite sure what you mean by this question. Could you elaborate?

I understand from the procedure to serialize a CSS value that it expects component value(s) to be used. The current definition of a component value does not include a string, but only preserverd tokens, function, and block. And I understand from the above comments that a `var()`-containg value will have to be serialized using the original string, as is already the case of a custom property value.

Therefore I was wondering if a component value can be string. But I realize that I can just understand that these cases (custom property and `var()`-containing value) should ignore what the serialization procedure defines.

**Comment by emilio** [comment]
> Then the trailing whitespace should be removed when serializing color:__var(__--blue__)__, isn't it? The output in Firefox is color: var(__--blue__)__.

Yeah, I think that is a Gecko bug, mind filing it?

**Comment by tabatkins** [comment]
`<string-token>` is a kind of component value, sure. But the subject at hand is "the original string", aka the actual character data pre-tokenization. This is a minor layering violation; it requires reaching back into the original stylesheet text (or carrying that stylesheet text fragment forward). The specs (intentionally) don't define *precisely* how this is done.

**Comment by cdoublev** [comment]
> Right, as per those issues you linked in your edit, [leading (and trailing) whitespace is trimmed by the Syntax spec](https://drafts.csswg.org/css-syntax/#consume-declaration) before anything else has a chance to see the value [...].

`style.setProperty(prop, value)` (or `style[prop] = value`) does not seem to run **consume a declaration** (the linked procedure), which seems to be only used for `style.cssText = 'prop: value'`, or any other procedure that would trim trailing whitespaces. Do they need to be trimmed for both interfaces?

```js
e1.style.color = 'var(--color)  '
e2.style = 'color: var(--color)  ;'
e1.style.color === e2.style.color // ?
```

Am I missing something or is it missing from [parse a list of component values](https://drafts.csswg.org/css-syntax-3/#parse-a-list-of-component-values)?

```diff
 1. Normalize input, and set input to the result.
- 2. Repeatedly consume a component value from input until an `<EOF-token>` is returned, appending the returned values (except the final `<EOF-token>`) into a list. Return the list.
+ 2. Repeatedly consume a component value from input until an `<EOF-token>` is returned, appending the returned values (except any trailing whitespaces and the final `<EOF-token>`) into a list. Return the list.

**Comment by cdoublev** [comment]
Please allow me to explain the issue more clearly:

**Case 1:** `style.cssText = 'color: var(--color) ;'`

1. [parse a CSS declaration block](https://drafts.csswg.org/cssom-1/#parse-a-css-declaration-block)
2. [parse a list of declarations](https://drafts.csswg.org/css-syntax-3/#parse-a-list-of-declarations)
3. [consume a list of declarations](https://drafts.csswg.org/css-syntax-3/#consume-a-list-of-declarations)
4. [consume a declaration](https://drafts.csswg.org/css-syntax-3/#consume-a-declaration)

Result: `var(--color)'`, ie. the trailing whitespace *is trimmed*.

**Case 2:** `style.color = 'var(--color) '` or (alias) `style.setProperty('color', 'var(--color) ')`

1. [parse a css value](https://drafts.csswg.org/cssom-1/#parse-a-css-value)
2. [parse a list of component values](https://drafts.csswg.org/css-syntax-3/#parse-a-list-of-component-values)
3. repeatedly [consume a component value](https://drafts.csswg.org/css-syntax-3/#consume-a-component-value)

Result: `var(--color) '`, ie. the trailing whitespace *is not trimmed*.

---

I was suggesting a change above to preserve the trailing whitespace but please note that this inconsistent behavior does not matter to me. Does it matter for any reason, eg. round-tripping? I'm just trying to match the (expected) browser outputs for both interfaces.

**Comment by cdoublev** [comment]
Also, and related to #2515: `style = 'padding: var(--p); padding-top: var(--p)'` serializes to `padding-right: ; padding-bottom: ; padding-left: ; padding-top: var(--p);` .

It could be fixed with:

```diff
- If all of the component longhand properties for a given shorthand are pending-substitution values from the same original shorthand value, the shorthand property must serialize to that original (var()-containing) value.
+ If all of the component longhand properties for a given shorthand are pending-substitution values equivalent to the original shorthand value, the shorthand property must serialize to that original (var()-containing) value.
```

Ie. *equivalent* would mean a deep comparison of the component value lists but ignoring comments.
